README.md
# neutralitytestr

### WARNING: It is recommended that you use [*mobster*](https://github.com/caravagn/mobster) for this type of analysis moving forward

[![Travis-CI Build Status](https://travis-ci.org/marcjwilliams1/neutralitytestr.svg?branch=master)](https://travis-ci.org/marcjwilliams1/neutralitytestr)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/neutralitytestr)](https://cran.r-project.org/package=neutralitytestr)
![](http://cranlogs.r-pkg.org/badges/grand-total/neutralitytestr)

This is an R package to analyse Variant Allele Frequency (VAF) distributions as reported from high throughput cancer sequencing. It reports 4 summary statistics and associated p-values based on a neutral model of tumour evolution, as well as functions to plot the VAF histogram and model fits.

## Getting Started
You can download the package from [CRAN](https://cran.r-project.org/package=neutralitytestr) in the usual way.
``` r
install.packages(neutralitytestr)
library(neutralitytestr)
```


To download the latest development version you'll need to use the ```devtools``` package:
```R
devtools::install_github("marcjwilliams1/neutralitytestr")
```

## Analysis
The package comes with some preloaded test data, generated from a simulation of tumour growth. These test data sets are called ```VAFselection``` and ```VAFneutral```.

The basic functionality of the ```neutralitytestr``` package is achieved by creating a ```neutralitytest``` object. The ```neutralitytest``` object contains a range of metrics to test for neutrality, and makes plotting histograms and cumulative distributions to visualize the output easy. The ```neutralitytest``` function takes a vector of VAFs and an upper and lower limit for the frequency range over which we wish to test whether the data is consistent with a neutral model, and then calculates all 4 metrics.
```R
out <- neutralitytest(VAFneutral, fmin = 0.1, fmax = 0.25)
```

The neutralitytest object can be summarised using the ```summary(out)``` command.
```
Summary of neutrality metrics:

Area:
  value =  0.03067203 , p-value =  0.679
Kolmogorov Distance:
  value =  0.09036603 , p-value =  0.641
Mean distance:
  value =  0.04131414 , p-value =  0.595
R^2:
  value =  0.98993 , p-value =  0.404

Effective mutation rate =  216.7985
```


The following commands will plot a VAF histogram, a least squares model fit and the normalized distributions. For more information see the vignette.

```R
vaf_histogram(out)
lsq_plot(out)
normalized_plot(out)
```

We can also input the read depth, cellularity, overdispersion rho and ploidy and let the package calculate an appropriate upper integration limit by considering the expected standard deviation of the clonal peak. Using this on the VAFselection data we would do the following:

```R
out <- neutralitytest(VAFselection, read_depth = 100.0, cellularity = 0.8, rho = 0.0, ploidy = 2)
plot_all(out) #this will plot all 3 of the above plots and combine into 1 figure.
```

For a more detailed introduction to the package see this [vignette](https://cran.r-project.org/web/packages/neutralitytestr/vignettes/neutraltytestr.html).

### Notes
Note that the p-values should be interpreted with care and are meant to serve as an approximation to guide the interpretation of the test statistics. These p-values were generated empirically from a simulated cohort of cancers with known ground truth and are derived from the same data that generated the ROC curves in supplementary figure 3 from the paper. This cohort of simulated tumours were "sequenced" to 100X and thus if a sample you are analysing is sequenced to much higher or lower depth the p-values may no longer be valid. We have also developed a Bayesian alternative to identifying neutral and non-neutral tumours, this is available [here](https://github.com/marcjwilliams1/SubClonalSelection.jl). Note that this Bayesian method is much more computationally expensive and can take upwards of 10 hours per sample.


NEWS.md
# neutralitytestr 0.0.3
- Add warning message notifying users that the package has been superseded by another package called mobster.

# neutralitytestr 0.0.2
- Added option so that maximum of integration range can be calculated from read read depth
- Added mu/beta = annotation to lsq_plot rather than y = 

# neutralitytestr 0.0.1

First realease of package.


R/methods.R
#' Testing for neutrality on cancer sequencing data
#'
#' \code{neutralitytest} returns a neutralitytest object which contains the result of
#' various test statistics to test for neutrality as described in Williams et al. Nature Genetics 2018.
#' WARNING: This package has been superseded by MOBSTER, see Caravagna et al. Nature Genetics 2020.
#'
#'
#' @param VAF Vector of variant allele frequencies (VAFs) from a deep sequencing experiment,
#' numbers should be between 0 and 1
#' @param fmin Minimum VAF of integration range, default is 0.1
#' @param fmax Maximum VAF of integration range, default is 0.25
#' @param read_depth Read depth of sample, if this is specified it will be used to calculate an appropriate integration range.
#' default is NULL in which case the default or inputted fmin and fmax will be used.
#' @param rho Overdispersion of sample if known, default is 0.0. Will be used to calculate integration range if read_depth != NULL
#' @param cellularity Cellularity of sample, default is 1.0. Will be used to calculate integration range if read_depth != NULL
#' @param ploidy Ploidy of the genome, default is 2. Ideally mutations should be filtered for this ploidy before running the test.
#' Will be used to calculate integration range if read_depth != NULL
#' @return neutralitytest object which contains test statistics which tests
#' if the sequencing data is consistent a neutral evolutionary model.
#' Test statistics are area between theoretical and empirical curves, kolmogorov distance, mean distance and R^2 statistics
#' from linear model fit. Also returns an estimate of the mutation rate per tumour tumour doubling, the raw VAFs and
#' cumulative distribution
#' @examples
#' neutralitytest(runif(100))
#' neutralitytest(VAFselection, fmin = 0.1, fmax = 0.25)
#' neutralitytest(VAFneutral, read_depth = 100.0, cellularity = 0.8)
#' @export
neutralitytest <- function(VAF, fmin = 0.1, fmax = 0.25, read_depth = NULL, rho = 0.0, cellularity = 1.0, ploidy = 2) {

  message("This package has been superseded by MOBSTER, see Caravagna et al. Nature Genetics 2020.")

  if (!is.null(read_depth)){
    message("Using inputted read depth to calculate integration range")
    SD <- sqrt((1 + (read_depth - 1) * rho) / read_depth)
    clonalpeak <- cellularity / ploidy
    fmax <- clonalpeak - 2*SD
    fmax <- round(fmax, 2)
    message(paste0(" Standard deviation of clonal peak is calculated to be ", SD, "\n Mean of peak is ", clonalpeak,
                   ". \n Using integration range - (", fmin, ",",fmax, ")"))
  }
  cumulativefrequency <- analyse_vaf(VAF, fmin, fmax)

  A <- areametric(cumulativefrequency, fmin, fmax)

  Dk <- kolmogorovdist(data.frame(VAF = VAF), fmin, fmax)

  meanD <- meandist(cumulativefrequency)

  fit <- lsqfit(cumulativefrequency, fmin, fmax)

  out <- list(mutation.rate = fit$mu,
              model = fit,
              rsq = data.frame(metric = fit$rsq, pval = fit$pval),
              area = A,
              Dk = Dk,
              meanDist = meanD,
              cumulativefrequency = cumulativefrequency,
              VAF = VAF)

  class(out) <- "neutralitytest"

  invisible(out)

}

#' @export
print.neutralitytest <- function(x, ...){
  cat("Summary of neutrality metrics:","\n\n")

  cat("Area:\n ","value = ", x$area$metric, ", p-value = ", x$area$pval,"\n")
  cat("Kolmogorov Distance:\n ","value = ", x$Dk$metric, ", p-value = ", x$Dk$pval,"\n")
  cat("Mean distance:\n ","value = ", x$meanD$metric, ", p-value = ", x$meanD$pval,"\n")
  cat("R^2:\n ","value = ", x$rsq$metric, ", p-value = ", x$rsq$pval,"\n\n")

  cat("Effective mutation rate = ",x$mutation.rate)
}

#' @export
summary.neutralitytest <- function(object, ...){
  cat("Summary of neutrality metrics:","\n\n")

  cat("Area:\n ","value = ", object$area$metric, ", p-value = ", object$area$pval,"\n")
  cat("Kolmogorov Distance:\n ","value = ", object$Dk$metric, ", p-value = ", object$Dk$pval,"\n")
  cat("Mean distance:\n ","value = ", object$meanD$metric, ", p-value = ", object$meanD$pval,"\n")
  cat("R^2:\n ","value = ", object$rsq$metric, ", p-value = ", object$rsq$pval,"\n\n")

  cat("Effective mutation rate = ", object$mutation.rate)
}
